### PR TITLE
Serialize EC private keys

### PR DESCRIPTION
**Pull request by reaperhulk**

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/ec.py [hunk 1]
```diff
@@ -200,6 +200,23 @@ def private_numbers(self):
             public_numbers=self.public_key().public_numbers()
         )
 
+    def private_bytes(self, encoding, format, encryption_algorithm):
+        evp_pkey = self._backend._lib.EVP_PKEY_new()
+        assert evp_pkey != self._backend._ffi.NULL
+        evp_pkey = self._backend._ffi.gc(
+            evp_pkey, self._backend._lib.EVP_PKEY_free
+        )
+        res = self._backend._lib.EVP_PKEY_set1_EC_KEY(evp_pkey, self._ec_key)
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/ec.py]: Ugh, why do we have an `EVP_PKEY` on RSA but not EC?

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/ec.py]: This is a question dreid asked awhile back in an issue I can't find at the moment. The short answer is "we didn't need it until now". We can hoist it up into the init now though if we want.

Comment (alex) [inline src/cryptography/hazmat/backends/openssl/ec.py]: I'd kind of like to see what that patch looks like if you don't mind

**Comment by alex** [comment]
This LGTM, as with most substantitve things, I'd love another pair of eyes (cc: @public).

File: src/cryptography/hazmat/backends/openssl/ec.py (hunk 1 above)
Comment (public) [inline src/cryptography/hazmat/backends/openssl/ec.py]: I found that the EVP APIs get in the way and are often less obvious than the low level APIs for the kind of stuff we do but maybe I was insufficiently good at OpenSSL back then.

Comment (alex) [inline src/cryptography/hazmat/backends/openssl/ec.py]: I don't think we have a choice about using EVP for PKCS8.

On Mon, Mar 2, 2015 at 7:39 AM, Alex Stapleton <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/ec.py
> <https://github.com/pyca/cryptography/pull/1709#discussion_r25607137>:
>
> > @@ -200,6 +200,23 @@ def private_numbers(self):
> >              public_numbers=self.public_key().public_numbers()
> >          )
> >
> > +    def private_bytes(self, encoding, format, encryption_algorithm):
> > +        evp_pkey = self._backend._lib.EVP_PKEY_new()
> > +        assert evp_pkey != self._backend._ffi.NULL
> > +        evp_pkey = self._backend._ffi.gc(
> > +            evp_pkey, self._backend._lib.EVP_PKEY_free
> > +        )
> > +        res = self._backend._lib.EVP_PKEY_set1_EC_KEY(evp_pkey, self._ec_key)
>
> I found that the EVP APIs get in the way and are often less obvious than
> the low level APIs for the kind of stuff we do but maybe I was
> insufficiently good at OpenSSL back then.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1709/files#r25607137>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (public) [inline src/cryptography/hazmat/backends/openssl/ec.py]: Using the OpenSSL API, maybe not, but I think we generally do need access to the low level key objects for the other operations we implement.

**Comment by public** [comment]
LGTM

**Comment by alex** [comment]
Needs a CHANGELOG and then I'll merge

**Pull request merged**